pkg/k8sapi/discovery.go

```diff
@@ -185,6 +185,11 @@ func (d *Controller) K8SGetLocalPodIPs() ([]*K8SPodInfo, error) {
 	return localPods, nil
 }
 
+func (d *Controller) WaitUntilReady() {
+	stopCh := make(chan struct{})
+	cache.WaitForCacheSync(stopCh, d.controller.informer.HasSynced)
```

Comment:
What about the `Controller.synced` field? In `Controller.run`, the `synced` field is set to true:

```go
	go d.controller.informer.Run(stopCh)
	log.Info("Waiting for controller cache sync")
	// Wait for all involved caches to be synced, before processing items from the queue is started
	if !cache.WaitForCacheSync(stopCh, d.controller.informer.HasSynced) {
		log.Error("Timed out waiting for caches to sync!")
		runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
		return
	}

	log.Info("Synced successfully with APIServer")
	d.synced = true
```

It seems to me that the simplest solution to this problem is to constrain all this setup/cache-prep logic to the construction of the `Controller` object itself by creating the `SharedInformer` structs in the `NewController` function and then returning the ready-to-run `Controller` object. You could remove the `Controller.synced` field entirely if you did this, and the code would become a whole lot easier to read.